src/DataLayouts/struct.jl

```diff
@@ -11,46 +11,92 @@
-    basetype(ntuple(i -> fieldtype(S, i), fieldcount(S))...)
-end
-function basetype(::Type{S1}, Sx...) where {S1}
-    if sizeof(S1) == 0
-        return basetype(Sx...)
+is_valid_basetype(::Type{T}, ::Type{S}) where {T, S} =
+    sizeof(S) == 0 ||
+    (fieldcount(S) > 0 && is_valid_basetype(T, fieldtypes(S)...))
+is_valid_basetype(::Type{T}, ::Type{<:T}) where {T} = true
+is_valid_basetype(::Type{T}, ::Type{S}, Ss...) where {T, S} =
+    is_valid_basetype(T, S) && is_valid_basetype(T, Ss...)
+
+"""
+    check_basetype(::Type{T}, ::Type{S})
+
+Check whether the types `T` and `S` have well-defined sizes, and whether an
+object of type `S` can be stored in a container with elements of type `T`.
+"""
+function check_basetype(::Type{T}, ::Type{S}) where {T, S}
+    if @generated
+        if !isbitstype(T)
+            estr = "Container type $T has indeterminate size"
+            :(error($estr))
+        elseif !isbitstype(S)
+            estr = "Element type $S has indeterminate size"
+            :(error($estr))
+        elseif !is_valid_basetype(T, S)
+            estr = "Element type $S cannot be stored in container type $T"
+            :(error($estr))
+        else
+            :(return nothing)
+        end
+    else
+        isbitstype(T) || error("Container type $T has indeterminate size")
+        isbitstype(S) || error("Element type $S has indeterminate size")
+        is_valid_basetype(T, S) ||
+            error("Element type $S cannot be stored in container type $T")
+        return nothing
     end
-    FT1 = basetype(S1)
-    FT2 = basetype(Sx...)
-    FT1 !== FT2 && error("Inconsistent basetypes $FT1 and $FT2")
-    return FT1
 end
 
-replace_basetype(::Type{S}, ::Type{FT}) where {S <: AbstractFloat, FT} = FT
-function replace_basetype(::Type{S}, ::Type{FT}) where {S <: Tuple, FT}
-    Tuple{ntuple(i -> replace_basetype(fieldtype(S, i), FT), fieldcount(S))...}
-end
-function replace_basetype(
-    ::Type{NamedTuple{names, T}},
-    ::Type{FT},
-) where {names, T, FT}
-    NamedTuple{names, replace_basetype(T, FT)}
-end
+"""
+    replace_basetype(::Type{T}, ::Type{T′}, ::Type{S})
+
+Changes the type parameters of `S` to produce a new type `S′` such that, if
+`is_valid_basetype(T, S)` is true, then so is `is_valid_basetype(T′, S′)`.
+"""
+replace_basetype(::Type{T}, ::Type{T′}, ::Type{S}) where {T, T′, S} =
+    length(S.parameters) == 0 ? S :
+    S.name.wrapper{replace_basetypes(T, T′, S.parameters...)...}
+replace_basetype(::Type{T}, ::Type{T′}, ::Type{<:T}) where {T, T′} = T′
+replace_basetype(::Type{T}, ::Type{T′}, value) where {T, T′} = value
+replace_basetypes(::Type{T}, ::Type{T′}, value) where {T, T′} =
+    (replace_basetype(T, T′, value),)
+replace_basetypes(::Type{T}, ::Type{T′}, value, values...) where {T, T′} =
+    (replace_basetype(T, T′, value), replace_basetypes(T, T′, values...)...)
+# TODO: This could potentially lead to some annoying bugs, since it replaces
```

Comment:
yeah there are probably better ways (or at least more checks we could add) to do this but this is a good writeup of the issues